### PR TITLE
Time limit for cache sync on startup and controller reconciliations

### LINKED ISSUE #99: Worker routines get stuck
**How to categorize this issue?**

/area robustness
/kind bug
/priority normal

**What happened**:

We have observed some situations, were grm gets stuck reconciling a specific managed resource and does not act upon it anymore.
In all cases I observed, it was either happening in conjunction with a longer period of downtime of the source or target API server (before #95) or a large amount of secret data in the target cluster (like described in #92).

**What you expected to happen**:

grm should not get stuck and reconcile all managed resources with the given sync interval.

**How to reproduce it (as minimally and precisely as possible)**:

Not sure yet.
My guess would be that the worker goroutines get stuck in some `WaitForCacheSync`, when the API server is unavailable for a longer period of time or the amount of watched data is to big.

**Anything else we need to know?**:

**Environment**:

- Gardener-Resource-Manager version: v0.20.0
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Others:

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area scalability
/kind bug
/priority normal

**What this PR does / why we need it**:
In order to prevent stuck worker routines as described in #99, @timebertt suggested to introduce time limits for the reconciliations of the controllers, similarly to how it's already done in Gardener's shoot care controller ([see](https://github.com/gardener/gardener/blob/efd33e4e97656127e8bed12a152be4e250604599/pkg/gardenlet/controller/shoot/shoot_care_control.go#L221-L222)).

Also, the cache sync on startup is now limited to `5m`.

**Which issue(s) this PR fixes**:
Fixes #99

**Release note**:

```noteworthy operator
The controller reconciliations are now limited to `1m`.
```

### CODE REVIEW

**Comment by timebertt** [comment]
```
>> make generate
make generate needs to be run:
 M api/resources/v1alpha1/zz_generated.deepcopy.go
```

File: cmd/gardener-resource-manager/app/app.go [hunk 1]
```diff
@@ -113,7 +114,10 @@ func NewResourceManagerCommand() *cobra.Command {
 				}
 			}()
 
-			if !targetClientOpts.Completed().WaitForCacheSync(ctx.Done()) {
+			ctxWaitForCache, cancelWaitForCache := context.WithTimeout(ctx, 5*time.Minute)
```
Comment (timebertt) [inline cmd/gardener-resource-manager/app/app.go:117]: Side-note, doesn't have to be tackled in this PR: we should further improve health checks, by separating liveness and readiness endpoint and report an unready state as long as the caches haven't synced initially. 
Though, this is also a general problem in our components.

**Review by timebertt — APPROVED** [review]
Very nice, thank you!
This was exactly how I imagined it 🚀 
/lgtm

**Comment by rfranzke** [comment]
Due to the `go mod` magic I needed to run `make revendor` after the `make generate`. :( :)

**Review by timebertt — APPROVED** [review]

**Pull request merged**